include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -2055,6 +2055,200 @@ __parallel_reduce_by_segment(oneapi::dpl::__internal::__device_backend_tag, _Exe
         oneapi::dpl::unseq_backend::__has_known_identity<_BinaryOperator, __val_type>{});
 }
 
+//------------------------------------------------------------------------
+// parallel_scan_by_segment - sync pattern
+//------------------------------------------------------------------------
+template <typename _CustomName, bool __is_inclusive, typename _Range1, typename _Range2, typename _Range3,
+          typename _BinaryPredicate, typename _BinaryOperator, typename _InitType>
+auto
+__parallel_scan_by_segment_reduce_then_scan(sycl::queue& __q, _Range1&& __keys, _Range2&& __values,
+                                            _Range3&& __out_values, _BinaryPredicate __binary_pred,
+                                            _BinaryOperator __binary_op, [[maybe_unused]] _InitType __init)
+{
+    using _GenReduceInput = __gen_scan_by_seg_reduce_input<_BinaryPredicate>;
+    using _ReduceOp = __scan_by_seg_op<_BinaryOperator>;
+    using _GenScanInput = __gen_scan_by_seg_scan_input<_BinaryPredicate>;
+    using _ScanInputTransform = __get_zeroth_element;
+    using _ValueType = oneapi::dpl::__internal::__value_t<_Range2>;
+    std::size_t __n = __keys.size();
+    // TODO: A bool type may be used here for a smaller footprint in registers / temp storage but results in IGC crashes
+    // during JIT time. The same occurs for uint8_t and uint16_t. uint32_t is used as a workaround until the underlying
+    // issue is resolved.
+    using _FlagType = std::uint32_t;
+    if constexpr (__is_inclusive)
+    {
+        oneapi::dpl::unseq_backend::__no_init_value<oneapi::dpl::__internal::tuple<_FlagType, _ValueType>>
+            __wrapped_init;
+        using _WriteOp = __write_scan_by_seg<__is_inclusive, decltype(__wrapped_init), _BinaryOperator>;
+        return __parallel_transform_reduce_then_scan<sizeof(oneapi::dpl::__internal::tuple<_FlagType, _ValueType>),
+                                                     _CustomName>(
+            __q, __n,
+            oneapi::dpl::__ranges::make_zip_view(std::forward<_Range1>(__keys), std::forward<_Range2>(__values)),
+            std::forward<_Range3>(__out_values), _GenReduceInput{__binary_pred}, _ReduceOp{__binary_op},
+            _GenScanInput{}, _ScanInputTransform{}, _WriteOp{__wrapped_init, __binary_op}, __wrapped_init,
+            /*Inclusive*/ std::true_type{}, /*_IsUniquePattern=*/std::false_type{});
+    }
+    else
+    {
+        // The init value is manually applied through the write functor in exclusive-scan-by-segment and we always pass __no_init_value.
+        // This is because init handling must occur on a per-segment basis and functions differently than the typical scan init.
+        oneapi::dpl::unseq_backend::__init_value<oneapi::dpl::__internal::tuple<_FlagType, _ValueType>> __wrapped_init{
+            {0, __init.__value}};
+        using _WriteOp = __write_scan_by_seg<__is_inclusive, decltype(__wrapped_init), _BinaryOperator>;
```

Comment:
Just a preference thing, but it might be cleaner to define a type alias for `oneapi::dpl::unseq_backend::__init_value<oneapi::dpl::__internal::tuple<_FlagType, _ValueType>>` and use that both in the declaration above this line and in this template.

Reply:
Good point, I combined this with Sergey's suggestion to move it outside of the branch and added an alias for the tuple type since it is used elsewhere as well.